### PR TITLE
clientv3: start a session with existing lease

### LINKED ISSUE #7166: Start a Session with existing lease
A client uses a session for election, becomes the leader and then restarts. Would it be possible for the client to reclaim the leader role on restart? Assume the client can figure out the existing lease id after restart. If the session can be restarted with existing lease id, the client can maintain leadership for that election.

cc @xiang90

### PR DESCRIPTION
**Pull request by ravigadde**
This change is needed to handle process restarts with elections. When the
leader process is restarted, it should be able to hang on to the leadership
by using the existing lease.

Fixes #7166

### CODE REVIEW

**Comment by ravigadde** [comment]
cc @xiang90

File: clientv3/concurrency/session.go [hunk 1]
```diff
@@ -115,6 +119,15 @@ func WithTTL(ttl int) SessionOption {
 	}
 }
 
+// WithExistingLease specifies the existing leaseID to be used for the session.
+// This is useful in process restart scenario, for example, to reclaim
+// leadership from an election prior to restart.
+func WithExistingLease(leaseID v3.LeaseID) SessionOption {
```
Comment (heyitsanthony) [inline clientv3/concurrency/session.go]: just  `WithLease`?

**Comment by ravigadde** [comment]
Are integration tests working? I added a case, but see this failure:
$ PASSES="integration" ./test
Running integration tests...
...
=== RUN   TestDoubleClusterSizeOf3
panic: test timed out after 15m0s

**Comment by xiang90** [comment]
@ravigadde 

The timeout is fixed on master I think. 

But for simplicity, try to do something like

```
cd $ETCD/clientv3/integration
go test -run=$YOUR_NEW_TEST
```

**Comment by ravigadde** [comment]
@xiang90 @heyitsanthony PTAL, updated the PR. Jenkins failure is because of this:

--- FAIL: TestRepairWriteTearMiddle (0.01s)
	repair_test.go:87: fix = false, want true
2017-01-30 13:55:35.378827 E | wal: failed to allocate space when creating new wal file (no space left on device)
2017-01-30 13:55:35.388656 E | wal: failed to allocate space when creating new wal file (no space left on device)

File: integration/v3_election_test.go [hunk 2]
```diff
@@ -225,3 +225,39 @@ func TestElectionOnPrefixOfExistingKey(t *testing.T) {
...
+	defer session2.Orphan()
+
+	e2 := concurrency.NewElection(session2, "test-elect")
+	if err3 := e2.Campaign(context.TODO(), "def"); err3 != nil {
+		t.Fatal(err3)
+	}
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	if resp := <-e.Observe(ctx); len(resp.Kvs) == 0 || string(resp.Kvs[0].Value) != "def" {
```
Comment (xiang90) [inline integration/v3_election_test.go]: can we add a timeout for this?

**Comment by xiang90** [comment]
One minor issue. LGTM. Defer to @heyitsanthony

**Comment by ravigadde** [comment]
Thanks @xiang90 updated the PR

File: integration/v3_election_test.go [hunk 3]
```diff
@@ -225,3 +225,39 @@ func TestElectionOnPrefixOfExistingKey(t *testing.T) {
...
+// in a new session with the same lease id) does not result in loss of
+// leadership.
+func TestElectionOnSessionRestart(t *testing.T) {
+	clus := NewClusterV3(t, &ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+	cli := clus.RandClient()
+
+	session, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
```
Comment (xiang90) [inline integration/v3_election_test.go]: for the setup stage, we should use t.Fatal. for the actual checking stage, we should use t.Error.

i will comment below as well.

sorry for not catching this up at the beginning.

File: integration/v3_election_test.go [hunk 4]
```diff
@@ -225,3 +225,39 @@ func TestElectionOnPrefixOfExistingKey(t *testing.T) {
...
+
+	e2 := concurrency.NewElection(session2, "test-elect")
+	if err3 := e2.Campaign(context.TODO(), "def"); err3 != nil {
+		t.Fatal(err3)
+	}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 5*time.Second)
+	defer cancel()
+	if resp := <-e2.Observe(ctx); len(resp.Kvs) == 0 || string(resp.Kvs[0].Value) != "def" {
+		t.Fatalf("expected value=%q, got response %v", "def", resp)
```
Comment (xiang90) [inline integration/v3_election_test.go]: this should be t.Errorf

File: integration/v3_election_test.go [hunk 5]
```diff
@@ -225,3 +225,39 @@ func TestElectionOnPrefixOfExistingKey(t *testing.T) {
...
+	}
+
+	e := concurrency.NewElection(session, "test-elect")
+	if err1 := e.Campaign(context.TODO(), "abc"); err1 != nil {
+		t.Fatal(err1)
+	}
+
+	session2, err2 := concurrency.NewSession(cli, concurrency.WithLease(session.Lease()))
+	if err2 != nil {
+		t.Error(err2)
```
Comment (xiang90) [inline integration/v3_election_test.go]: t.fatal

Comment (ravigadde) [inline integration/v3_election_test.go]: Done

File: integration/v3_election_test.go [hunk 6]
```diff
@@ -225,3 +225,39 @@ func TestElectionOnPrefixOfExistingKey(t *testing.T) {
...
+	session, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	e := concurrency.NewElection(session, "test-elect")
+	if err1 := e.Campaign(context.TODO(), "abc"); err1 != nil {
+		t.Fatal(err1)
+	}
+
```
Comment (heyitsanthony) [inline integration/v3_election_test.go]: can this also test that a waiter is never elected on the failover by having a waiter right here?

```go
// ensure leader is not lost to waiter on fail-over
waitSession, werr := concurrency.NewSession(cli)
if werr != nil {
    t.Fatal(werr)
}
defer waitSession.Orphan()
waitCtx, waitCancel := context.WithCancel(context.TODO())
defer waitCancel()
go concurrency.NewElection(waitSession, "test-elect").Campaign(waitCtx, "123")
```

Comment (ravigadde) [inline integration/v3_election_test.go]: Sure, added this code.

**Comment by heyitsanthony** [comment]
lgtm

**Comment by xiang90** [comment]
LGTM

**Pull request merged**